compressor: expose generic compressor filter to users

rojkov opened the pull request:
Description:

Currently the generic HTTP compressor filter isn't exposed to users
even though it's used internally by `envoy.filters.http.gzip` and can be
used by external filter extensions.

Expose the compressor's config API to users. For example the filter
can be configured as follows:
```yaml
    ...
    filter_chains:
      filters:
      - name: envoy.http_connection_manager
        config:
           http_filters:
           - name: envoy.filters.http.compressor
             config:
               disable_on_etag_header: true
               content_length: 100
               content_type:
                 - text/html
                 - application/json
               compressor_library:
                 name: text_optimized
                 typed_config:
                   "@type": type.googleapis.com/envoy.extensions.filters.http.compressor.gzip.v3.Gzip
                   memory_level: 3
                   window_bits: 10
                   compression_level: best
                   compression_strategy: default_strategy
    ...
```
Multiple compressor filters using different compressor libraries,
e.g. gzip and brotli, can be stacked in one filter chain.

Risk Level: Medium
Testing: unit tests, manual testing
Docs Changes: a page for the new compressor filter is added, also comments in `compressor.proto` were updated.
Release Notes: added a link to compressor's doc to version_history.rst

rojkov commented:
/cc @junr03, @rebello95, @mattklein123

htuch commented on api/envoy/config/filter/http/compressor/gzip/v2/gzip.proto:
```diff
@@ -0,0 +1,62 @@
+syntax = "proto3";
```
How does this relate to https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/filter/http/gzip/v2/gzip.proto? Which one should service operators configure and when? Is the former deprecated?

mattklein123 replied:
Yes, exactly, the former is deprecated. We need to make that clear as @htuch points out, have text in deprecated.rst, and also make sure the deprecated feature warning/stat is incremented so that operators know thy need to upgrade.

htuch commented on docs/root/configuration/http/http_filters/compressor_filter.rst:
```diff
@@ -0,0 +1,80 @@
+
+Compressor
+==========
+Compressor is an HTTP filter which enables Envoy to compress dispatched data
```
Need links back and forth to relate to existing gzip filter, when one should be used vs. the other.

rojkov commented:
Thank you! I've just marked the old `gzip` filter explicitly deprecated.

But since it warns users now I wonder if we should revert https://github.com/envoyproxy/envoy/pull/10306 and make `gzip` use its old implementation.

htuch commented on api/envoy/config/filter/http/compressor/gzip/v2/gzip.proto:
```diff
@@ -0,0 +1,62 @@
+// [#next-free-field: 10]
+message Gzip {
+  enum CompressionStrategy {
+    DEFAULT = 0;
```
Can you comment on what the DEFAULTs are here and below? These should be fixed in the API.

htuch commented on api/envoy/config/filter/http/compressor/gzip/v2/gzip.proto:
```diff
@@ -0,0 +1,62 @@
+      BEST = 1;
+      SPEED = 2;
+    }
+  }
```
Why is the enum wraped with a message?

rojkov replied:
I presume this is because there is already `DEFAULT` value in `enum CompressionStrategy`. So the second `DEFAULT` was given its own namespace.
This format is copied from the current gzip filter.

Do you mind if I rename the items to reflect zlib's name directly? Like
```protobuf
message Gzip {
  enum CompressionStrategy {
    // corresponds to Z_DEFAULT_STRATEGY
    DEFAULT_STRATEGY = 0;
    // corresponds to Z_FILTERED
    FILTERED = 1; 
    // corresponds to Z_HUFFMAN_ONLY
    HUFFMAN_ONLY = 2;
    // corresponds to Z_RLE
    RLE = 3;
  }

  enum CompressionLevel {
    // corresponds to Z_DEFAULT_COMPRESSION
    DEFAULT_COMPRESSION = 0;
    // corresponds to Z_BEST_COMPRESSION 
    BEST_COMPRESSION = 1; 
    // corresponds to Z_BEST_SPEED
    BEST_SPEED = 2;
  }
...
}
```
Alternatively `compression_level` could be just an integer with valid values from 1 to 9 where 1 is the best compression (but slow), 9 is the best speed (but low compression) and the default is 6. Not as descriptive, but more flexible.

htuch replied:
The new structure you have looks great. Re: compression level, I think we should either have 1-9 integer or an enum that can grow to express this (ideally with enum values matching the corresponding gzip compression level numberings.

rojkov replied:
`Envoy::Compressor::ZlibCompressorImpl::CompressionLevel` is defined as enum, so I left this field  enum too, but extended with `COMPRESSION_LEVEL_X` values.

htuch commented on api/envoy/config/filter/http/compressor/v2/compressor.proto:
```diff
@@ -45,4 +50,25 @@ message Compressor {
+  // Compressor library specific configuration. See the supported libraries for further
+  // documentation.
+  oneof config_type {
+    google.protobuf.Struct config = 2 [deprecated = true];
```
No need to introduce untyped config for new APIs.

htuch commented on docs/root/configuration/http/http_filters/compressor_filter.rst:
```diff
@@ -0,0 +1,84 @@
+Compressor is an HTTP filter which enables Envoy to compress dispatched data
+from an upstream service upon client request. Compression is useful in
+situations where large payloads need to be transmitted without
+compromising the response time.
```
I think the tradeoff is more nuanced here; it has to do with latency, bandwidth and CPU tradeoff.

rojkov replied:
Updated this part. Please check.

htuch replied:
Nice, thanks, resolved.

mattklein123 commented on api/envoy/config/filter/http/compressor/v2/compressor.proto:
```diff
@@ -45,4 +50,25 @@ message Compressor {
+  //
+  // This field is ignored when used in the context of the deprecated "envoy.filters.http.gzip"
+  // filter.
+  CompressorLibrary compressor_library = 6;
```
I'm not sure it makes sense to have a default here. Can we require this and then make the gzip config explicit for clarity?

mattklein123 replied:
Can we make it required via annotation?

rojkov replied:
Hm... This message is used also in the deprecated `envoy.filters.http.gzip` filter. I guess there are two options:

1. revert #10306 completely or partially (drop the `compressor` field from the `Gzip` message);
2. duplicate `compressor.proto` for `envoy.filters.http.gzip`.

mattklein123 commented on docs/root/configuration/http/http_filters/compressor_filter.rst:
```diff
@@ -0,0 +1,84 @@
+.. _config_http_filters_compressor:
+
+Compressor
```
Please provide an example config that explicitly configures the gzip compressor sub-type. The nested types/configs are very confusing for users.

mattklein123 commented on api/envoy/config/filter/http/compressor/v2/compressor.proto:
```diff
@@ -45,4 +49,21 @@ message Compressor {
+}
+
+message CompressorLibrary {
+  // The name of the compressor library to instantiate for the filter.
+  string name = 1 [(validate.rules).string = {min_bytes: 1}];
```
We can remove this for new extensions as the extension will be looked up via the typed_config.

rojkov replied:
Yeap, removed.

rojkov commented:
@htuch I tried to follow the proto adding steps, but now `proto_format.sh fix` fails for me with
```
Traceback (most recent call last):
  File "/home/rojkov/work/envoy/bazel-bin/tools/protoxform/merge_active_shadow.runfiles/envoy/tools/protoxform/merge_active_shadow.py", line 142, in <module>
    pathlib.Path(dst).write_text(str(MergeActiveShadowFile(active_proto, shadow_proto)))
  File "/home/rojkov/work/envoy/bazel-bin/tools/protoxform/merge_active_shadow.runfiles/envoy/tools/protoxform/merge_active_shadow.py", line 122, in MergeActiveShadowFile
    MergeActiveShadowMessage(msg, shadow_msgs[msg.name], target_file_proto.message_type.add())
KeyError: 'CompressorLibrary'
```

I also removed ` option (udpa.annotations.versioning).previous_message_type = "envoy.config.filter.http.compressor.v2.CompressorLibrary";` from the message. But it didn't help.

htuch commented:
@rojkov can you post your WiP branch so I can try it out or look for the issue? Thanks.

rojkov commented:
@htuch sure, here it is https://github.com/rojkov/envoy/tree/compress-v2

htuch commented:
Thanks, that was helpful. There was a bug in the tooling, fix is in https://github.com/envoyproxy/envoy/pull/10696

mattklein123 commented:
Can you merge master and move the release notes to 1.15.0 and I will take another pass?

/wait

junr03 commented on source/extensions/filters/http/compressor/compressor_library_config.h:
```diff
@@ -0,0 +1,25 @@
+#pragma once
```
And as I think more about this I wonder if what we should have is extensions/compression/. And have the compression libraries be first class extensions.

And move the zlib implementation of the compressor there.

junr03 replied:
It makes sense to me, because we would like to be able to compile different compression libraries in and out. For example when we add brotli it will be beneficial to be able to compile out zlib.

junr03 commented on source/extensions/filters/http/compressor/factory_base.h:
```diff
@@ -0,0 +1,43 @@
+namespace Compressor {
+
+template <class ConfigProto>
+class CompressorLibraryFactoryBase : public NamedCompressorLibraryConfigFactory {
```
I am not quite seeing why we need the intermediary `NamedCompressorLibraryConfigFactory` class. I think we can just have the templated class derive directly from `Config::TypedFactory`

junr03 commented on source/extensions/filters/http/compressor/compressor_filter.cc:
```diff
@@ -0,0 +1,24 @@
+          compressor_factory->contentEncoding()),
+      compressor_factory_(std::move(compressor_factory)) {}
+
+std::unique_ptr<Envoy::Compressor::Compressor> CompressorFilterConfig::makeCompressor() {
```
qq: we were thinking about why we would lazily create the compressor (and thus have the config factory create a compressor factory) vs creating the compressor to begin with?

rojkov replied:
Compressors are created (or not) for every new stream if multiple conditions are met.

This particular wrapper method is going to get more meat when the `gzip` filter is dropped and the functionality of `Common::Compressors::CompressorFilterConfig` is moved to this class.

junr03 replied:
> Compressors are created (or not) for every new stream if multiple conditions are met.

Yep, this became clear as I read the rest of the PR. Sorry, I meant to delete this comment.

junr03 commented on source/extensions/filters/http/compressor/compressor_filter.cc:
```diff
@@ -0,0 +1,24 @@
+    const std::string& stats_prefix, Stats::Scope& scope, Runtime::Loader& runtime,
+    CompressorFactoryPtr compressor_factory)
+    : Common::Compressors::CompressorFilterConfig(
+          generic_compressor, stats_prefix + compressor_factory->statsPrefix(), scope, runtime,
```
I think it would be valuable to insert "compressor" here e.g., `stats_prefix + "compressor." + compressor_factory->statsPrefix()`.

junr03 replied:
Actually in thinking more about it (in conjunction with my suggestion that compressors become an extension point) I think we should have three levels of namespaces:

1. The compressor filter stats
2. The general compressor stats
3. The specific compressor stats

For example, we might have something like:
```
http_filter.compressor_filter.response_compressed.counter
http_filter.compressor_filter.compressor.buffer_compressed_failure.counter
http_filter.compressor_filter.compressor.gzip.window_bits_exceeded.counter
```

rojkov replied:
What if there are two compressor filters configured to use different encodings for different content types? I guess `http_filter.compressor_filter.compressor.buffer_compressed_failure.counter` will become indistinguishable between the two. And `.compressor_filter.compressor.` sounds like a tautology.

junr03 replied:
Good point on the former, although that problem still holds with the current scheme. Perhaps we need an identifier per filter similar to listeners, there might already be something in place, let me take a look.

For the latter, I agree it is a tautology, but I still think there is value in the separate namespaces, one if for stats pertinent to the compressor filter and the other is for stats pertinent to the compressor the filter is using. Any suggestions on better naming? Or do you think it is overkill to give the compressor itself stats, and all the stats should live on the filter?

rojkov replied:
I'm trying to put on a user's hat. As a user I'd probably cared about a filter's stats only. If there're stats specific to a compressor they would live in `compression.compressor.*` and `compression.decompressor.*` namespaces.

junr03 replied:
Right, they would be under that, but because we are passing the stats_prefix they would be prefixed under the http filter. So we would know that those stats refer to the compressor used by the compressor http filter. So for instance:

`http_filter.compressor_filter.compression.gzip.compressor.window_bits_exceeded` would tell us about the times window_bit_exceeded happened as used by the http compressor filter. However, you are right that if two filters used the gzip compressor we could not discern between the two

rojkov replied:
Oops, missed.

Ok, I've moved compressor's stats to its own namespace. Now they look like
```
http.ingress_http.compressor.gzip.compressor.compressed: 3
http.ingress_http.compressor.gzip.compressor.content_length_too_small: 0
http.ingress_http.compressor.gzip.compressor.header_compressor_overshadowed: 0
http.ingress_http.compressor.gzip.compressor.header_compressor_used: 3
http.ingress_http.compressor.gzip.compressor.header_gzip: 3
http.ingress_http.compressor.gzip.compressor.header_identity: 0
http.ingress_http.compressor.gzip.compressor.header_not_valid: 0
http.ingress_http.compressor.gzip.compressor.header_wildcard: 0
http.ingress_http.compressor.gzip.compressor.no_accept_header: 0
http.ingress_http.compressor.gzip.compressor.not_compressed: 0
http.ingress_http.compressor.gzip.compressor.not_compressed_etag: 0
http.ingress_http.compressor.gzip.compressor.total_compressed_bytes: 624
http.ingress_http.compressor.gzip.compressor.total_uncompressed_bytes: 768

```

The first `compressor` relates a stat to the compressor filter and the second indicates compression direction: inflating or deflating.

junr03 replied:
Nice, so now a user knows that those are the compressor filter stats that is using a gzip compressor. My only remaining doubt is that if we put everything under that tree:
1. Where would stats owned by the compressor class go? If they go under the same tree then it would be hard to know which stats are coming from the filter and which are coming from the compressor itself at first glance. We could have the filter stats not have the second compressor prefix, and have stats exposed by the compressor itself be under the second compressor prefix:
```
http.ingress_http.compressor.gzip.compressed: 3
# I know this stat doesn't exist, nor any stats "owned" by the compressor itself, but what if there were?
http.ingress_http.compressor.gzip.compressor.window_bits_exceeded: 10
```
2. Do you think people might configure compressor filters in the same chain with the same compressor library but different parameters?

rojkov replied:
1. Right. We could remove the second `compressor` from `GzipCompressorFactory::statsPrefix()`'s return value in this PR. Then if compressors ever provide their own stats like `window_bits_exceeded` they'll need to add their own prefix (e.g. `compressor.`) to whatever prefix is given by the context (e.g. `http.ingress_http.compressor.gzip.`). WDYT?
2. Yes, this is natural that different Content-Types may require different compression strategies. After all the strategies themselves were introduced to improve compression for certain content types, i.e. RLE works better for PNG images.

junr03 replied:
1. sgtm!
2. @rojkov I mostly meant installing two compressor filters, both with the gzip compressor, but with different parameters. Because in that case the stats for both installations would still get mixed.

rojkov replied:
1. done :)
2. So did I. Consider this config
```
          http_filters:
          - typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.compressor.v3.Compressor
              content_type:
              - image/png
              compressor_library:
                typed_config:
                  "@type": type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip
                  compression_strategy: RLE
          - typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.compressor.v3.Compressor
              content_type:
              - text/html
              - application/javascript
              compressor_library:
                typed_config:
                  "@type": type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip
                  compression_strategy: DEFAULT_STRATEGY
```
Even though the same gzip compressor is used for both PNG and text contents it's configured to use different strategies for them as PNG benefits from RLE while HTML compression may underperform with it.

Perhaps we could add an optional `ID` field to compressor config to be added to the stats prefix just to distinguish different filters in stats. But so far users haven't requested it for the existing `gzip` filter.

junr03 replied:
Gotcha! Sorry I misunderstood strategy for using a different compression library altogether. Yes, that is the exact scenario I was thinking of.

We could use the filter's name field to add to the stat's tree to differentiate them. After https://github.com/envoyproxy/envoy/issues/9466 was closed the name can be an arbitrary string. So we could have:
```
http.ingress_http.compressor.png.gzip.
http.ingress_http.compressor.web.gzip.
```

wdyt @rojkov?

rojkov replied:
Yes, that sounds good. In this case suffixing with `gzip.` may be redundant even, unless no name is configured.
I presume this will require updating `createFilterFactoryFromProto()`'s signature to accommodate filter names. Right?

junr03 replied:
Hmm interesting point about using only the name if present, and the generic compression library's prefix otherwise. I still like the compression library name there, specially if we eventually have library-specific stats.

> I presume this will require updating createFilterFactoryFromProto()'s signature to accommodate filter names. Right?

Dang, for some reason I thought that we had access to the filter name in the factory function after #9587. Yep, that would be the way to do it. Originally I thought about only changing `createFilterFactoryFromProto` and storing the name, but the factory itself is static. I wanted to prevent changing the `createFilterFactoryFromProtoTyped()` signature to avoid churn in all filters, but I don't see a way around it... Unless we add a "name" field to the compressor filter config.

@htuch and @kyessenov what do you think? Now that the filter name can be arbitrary, we would like access to it in the compressor filter in order to distinguish stats trees between different installations of the filter with the same compression libraries.

junr03 commented on source/extensions/filters/http/compressor/gzip/config.cc:
```diff
@@ -0,0 +1,103 @@
+  }
+}
+
+uint64_t GzipCompressorFactory::memoryLevelUint(Protobuf::uint32 level) {
```
We have utilities to unwrap values with a default `PROTOBUF_GET_WRAPPED_OR_DEFAULT`

junr03 commented on source/extensions/filters/http/compressor/compressor_library_config.h:
```diff
@@ -0,0 +1,25 @@
+  ~NamedCompressorLibraryConfigFactory() override = default;
+
+  virtual CompressorFactoryPtr
+  createCompressorLibraryFromProto(const Protobuf::Message& config,
```
```suggestion
  createCompressorFactoryFromProto(const Protobuf::Message& config,
```

rojkov replied:
Thanks! Renamed.

rojkov commented on source/extensions/filters/http/compressor/compressor_factory.h:
```diff
@@ -0,0 +1,24 @@
+#pragma once
```
`CompressorFactory` provides `contentEncoding()` which is specific to HTTP. I had to introduce this factory as an intermediate step before the `gzip` filter is fully deprecated and dropped. When it's dropped we can merge the code in `source/extensions/filters/http/common` and `source/extensions/filters/http/compressor`. I've left a couple of TODO notes describing this intention.

rojkov commented on source/extensions/filters/http/compressor/compressor_factory.h:
```diff
@@ -0,0 +1,24 @@
+public:
+  virtual ~CompressorFactory() = default;
+
+  virtual std::unique_ptr<Envoy::Compressor::Compressor> createCompressor() PURE;
```
Thanks! Done.

rojkov commented:
@htuch docs are failing because I used a v2 type in deprecated `extensions/filters/http/gzip/v3/gzip.proto` (to avoid setting redundant `Compressor.compressor_library` in configs which is marked as required). Should I move to v4alpha then?

rojkov commented:
@junr03 Yes, I agree. This new tree makes sense to me. For example some people might want to compile out the gzip extension together with the zlib compressor/decompressor and to use [qatzip (HW accelerated gzip)](https://github.com/intel/kubernetes-qat-envoy/tree/master/envoy_qatzip/compressor) instead.

Do you want me to use the new structure in the scope of this PR?

junr03 commented:
> For example some people might want to compile out the gzip extension together with the zlib compressor/decompressor and to use [qatzip (HW accelerated gzip)](https://github.com/intel/kubernetes-qat-envoy/tree/master/envoy_qatzip/compressor) instead.

@rojkov exactly!

> Do you want me to use the new structure in the scope of this PR?

Yeah, that would be great, specially because it would be good to commit the new protos where we intend them to be.

I know that some things might not be in their final resting place because of the legacy gzip filter, but getting this PR as close to what we intend long-term would be good. We could even talk about bringing in the qatzip logic in as another compression library (in addition to brotli).

rojkov commented:
@junr03 I've restructured the tree a bit. Could you please review the chnages?

Probably `/include/envoy/compressor/compressor.h` and `/include/envoy/decompressor/decompressor.h` should also move under `/include/envoy/compression`?

junr03 commented:
> Probably `/include/envoy/compressor/compressor.h` and `/include/envoy/decompressor/decompressor.h` should also move under `/include/envoy/compression`?

Yep, that is what I did here https://github.com/envoyproxy/envoy/pull/10744/files#diff-34f5b808b9dbcb1797c724b880758c78

junr03 commented on include/envoy/compressor/compressor.h:
```diff
@@ -26,5 +26,7 @@ class Compressor {
   virtual void compress(Buffer::Instance& buffer, State state) PURE;
```
This file should also be moved to the new tree, right?

junr03 commented on source/extensions/compression/common/BUILD:
```diff
@@ -0,0 +1,19 @@
+
+envoy_package()
+
+envoy_cc_library(
```
What is your opinion about this being in common/ (as you did here) vs having a compressor/decompressor sub-tree (like I did in my [PR](https://github.com/envoyproxy/envoy/pull/10744/files#diff-2cec098aba17a8c67877d0772b1167a7))?

junr03 commented on source/common/compressor/zlib_compressor_impl.h:
```diff
@@ -30,24 +30,37 @@ class ZlibCompressorImpl : public Zlib::Base, public Compressor {
    * Enum values used to set compression level during initialization.
```
Why did this stay in this tree rather than out to the extension tree?

junr03 commented on test/extensions/compression/gzip/BUILD:
```diff
@@ -0,0 +1,34 @@
+envoy_package()
+
+envoy_extension_cc_test(
+    name = "gzip_compressor_library_test",
```
This should be under a compressor sub-tree like the prod code

junr03 commented on test/extensions/compression/gzip/BUILD:
```diff
@@ -0,0 +1,34 @@
+
+envoy_extension_cc_test(
+    name = "gzip_compressor_integration_test",
+    srcs = ["gzip_filter_test.cc"],
```
the gzip filter test should not be in this tree

rojkov commented:
@junr03 Thanks! I overlooked a lot of things. Hopefully I made them right this time. Please take another look.

junr03 commented on test/extensions/compression/gzip/compressor/BUILD:
```diff
@@ -16,8 +20,8 @@ envoy_cc_fuzz_test(
     deps = [
```
I'd put the fuzz test one directory up, because it is testing both compression and decompression.

junr03 commented on test/extensions/compression/gzip/compressor/gzip_compressor_library_test.cc:
```diff
@@ -0,0 +1,107 @@
+namespace Gzip {
+namespace Compressor {
+
+class GzipTest : public testing::Test {
```
Setting up tests that are friend classes of the class you are testing, and exposing private member variables is not very common in Envoy. I would err on collapsing these tests on the zlib_compressor_impl_test.cc file and test for the expectations via the public API.

rojkov replied:
Ok, merged the tests and dropped friend class from `GzipCompressorFactory`.

junr03 commented on test/extensions/filters/http/common/compressor/compressor_filter_test.cc:
```diff
@@ -22,8 +22,8 @@ namespace HttpFilters {
 
-class MockCompressor : public Compressor::Compressor {
-  void compress(Buffer::Instance&, ::Envoy::Compressor::State) override {}
+class MockCompressor : public Envoy::Compression::Compressor::Compressor {
```
This could be an actual mock, and the class below could be skipped. In addition per my comment above, I think we can clean the tests here that expose private functions.

Take a look at https://github.com/junr03/envoy/pull/1/files#diff-579860a53a9e1874e312dd0f7b2df943

rojkov replied:
I find it convenient to test private functions in isolation from each other. There's a TODO item though to move them in a separate class (using the pimpl idiom I presume).

junr03 replied:
nice, yeah I agree some of the private functions (for instance the ones that manipulate headers) could become functions in a static utility class that is shared between compressor and decompressor. We can do that in a subsequent PR.

junr03 commented on test/extensions/filters/http/compressor/gzip_compressor_filter_test.cc:
```diff
@@ -0,0 +1,170 @@
+namespace HttpFilters {
+namespace Compressor {
+
+class GzipCompressorFilterTest : public testing::Test {
```
I think that as long as we have good unit tests for the generic compressor filter and good unit tests for the gzip compressor, then we don't need unit tests for the gzip filter.

rojkov commented on include/envoy/compression/compressor/BUILD:
```diff
@@ -0,0 +1,35 @@
+)
+
+envoy_cc_library(
+    name = "compressor_factory_interface",
```
That's a micro-optimization for compilation time as there are two targets in source/extensions/filters/http/compressor/BUILD which depend on either one or another. Probably not worth it.

rojkov commented on source/common/common/BUILD:
```diff
@@ -420,16 +420,6 @@ envoy_cc_library(
     ],
 )
 
-envoy_cc_library(
```
Ok, reverted.

junr03 commented on test/mocks/compression/compressor/mocks.h:
```diff
@@ -0,0 +1,37 @@
+  MOCK_METHOD(void, compress, (Buffer::Instance & buffer, State state));
+};
+
+class MockCompressorFactory : public CompressorFactory {
```
I guess this is not currently used because the filter test still uses the old Config class which took a factory name rather than the factory pointer?

rojkov replied:
Not sure which Config class you're referring, but I've added a test case dedicated to `HttpFilters::Compressor::CompressorFilterConfig`. There this mock is used to check if a factory's methods are actually called.

junr03 replied:
Ah here, I see it now https://github.com/envoyproxy/envoy/pull/10553/files#diff-a905be3f5872395dc9391e1e82022f87R17

mattklein123 commented on CODEOWNERS:
```diff
@@ -20,6 +23,8 @@ extensions/filters/common/original_src @snowp @klarose
 # compressor used by http compression filters
 /*/extensions/filters/http/common/compressor @gsagula @rojkov @dio
+/*/extensions/filters/http/compressor @rojkov @dio
+/*/extensions/filters/http/compressor/gzip @gsagula @rojkov @dio
```
Can this and line 25 be removed?

rojkov replied:
Removed the line 27, but 25 is still needed according to `check_format.py fix`

mattklein123 commented on api/envoy/extensions/compression/common/compressor/v3/compressor.proto:
```diff
@@ -0,0 +1,22 @@
+option java_multiple_files = true;
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: CompressorLibrary]
```
nit: "Compressor Library"

mattklein123 commented on api/envoy/extensions/filters/http/gzip/v3/gzip.proto:
```diff
@@ -75,5 +75,37 @@ message Gzip {
   // the fields `content_length`, `content_type`, `disable_on_etag_header` and
   // `remove_accept_encoding_header` are ignored.
-  compressor.v3.Compressor compressor = 10;
+  Compressor compressor = 10;
```
Technically since we already froze v3, someone might be using this already. Is the local clone wire compatible with what came before? This is a little shady but I think I'm probably OK with this given how new v3 is. cc @envoyproxy/api-shepherds

rojkov replied:
Yes, the local clone is just a copy/paste from v2. I tried to use v2 first, and it even compiled and worked, but docs generating got broken.

htuch replied:
I'd prefer on principle not to break v3.. is it really so bad to keep the file where it was before?

rojkov replied:
If to leave it untouched then having these lines in config:
```
          ...
          http_filters:
          - name: envoy.filters.http.gzip
            config:
              compressor:
                content_length: 50
          - name: envoy.filters.http.router
            config: {}
```

causes Envoy fail to start with
```
Proto constraint validation failed (GzipValidationError.Compressor: ["embedded message failed validation"] | caused by CompressorValidationError.CompressorLibrary: ["value is required"])
```
Though in the context of the Gzip http-filter `CompressorLibrary` is hardcoded already: it's gzip and can't be anything else.

htuch replied:
Is it plausible this existing API is in use by somebody at this point? If so, we need to follow the API stability rules and not break for v3.

You might need to live with deprecating the existing scheme, which might be over-verbose and redundant, and introducing the new one side-by-side, if that is the situation.

rojkov replied:
@htuch The problem is not about being redundant: existing configurations will be broken, because this PR introduces a new mandatory field to `compressor.v3.Compressor` - `compressor_library`.
I guess the only option left is to update the generic filter to use v4alpha messages. Or maybe there's a better way?

htuch replied:
Maybe don't make it required and then have Envoy implementation code that can raise errors if missing for generic compressor?

rojkov replied:
Oh, right. I was too determent to make it required at the proto definition level.
Thanks! Now the frozen API isn't updated.

mattklein123 commented on source/extensions/filters/http/gzip/gzip_filter.cc:
```diff
@@ -72,14 +76,46 @@ uint64_t GzipFilterConfig::windowBitsUint(Protobuf::uint32 window_bits) {
 GzipFilterConfig::compressorConfig(const envoy::extensions::filters::http::gzip::v3::Gzip& gzip) {
+  envoy::extensions::filters::http::compressor::v3::Compressor compressor = {};
+
+  // The deprecated gzip filter is using a locally defined Compressor field to avoid setting the
+  // compressor_library field which is mandatory in v3. Here we convert gzip's Compressor to
+  // what is accepted by the generic compressor.
```
It's up to you, but you might consider doing a YAML/JSON round trip which might be easier to read and understand (see the protobuf message utlities in protobuf/utility.h. We do this in other places. As this is not perf critical it's probably fine.

rojkov replied:
Thank you! I didn't know about this utility. With the round trip the code is easier to read indeed.

junr03 commented:
@rojkov we are very close just missing closing two ([1](https://github.com/envoyproxy/envoy/pull/10553#discussion_r416241209), [2](https://github.com/envoyproxy/envoy/pull/10553#discussion_r416232411)) conversations! This is awesome work. Thanks for all the thoughtful discussion :)

junr03 commented:
@rojkov awesome work! Seems like you got @htuch's comment resolved. Any thoughts on https://github.com/envoyproxy/envoy/pull/10553#discussion_r417690591? 

> Alternatively, we could add a "name" or "stat_prefix" field to the compressor filter config, and use that.
The more I think about it, the more I like this option. It would be a pretty unobtrusive way to disambiguate stats trees from different compression filters without affecting the filter interface.

rojkov commented:
> The more I think about it, the more I like this option. It would be a pretty unobtrusive way to disambiguate stats trees from different compression filters without affecting the filter interface.

Agree. I like "stats_prefix" as more explicit, but would prefer it to be optional to keep minimal configs light. Are you ok with that?

rojkov commented:
> That sounds good to me. Include it if present.

Done!
I decided not to mandate a separator at the end of the prefix, e.g. one might want to have an underscore instead of a dot.

junr03 commented on api/envoy/extensions/filters/http/compressor/v3/compressor.proto:
```diff
@@ -46,4 +51,17 @@ message Compressor {
+  // Prefix added to the names of stats generated by this filter. It can be used to disambiguate stats
+  // trees from different compression filters. For example stats of a compressor filter limited to compress
+  // binaries blobs and tweaked accordingly can be prefixed with "bins.". And another compressor filter
+  // tweaked co compress text content only can be prefixed with "text.".
```
```suggestion
  // tweaked to compress text content only can be prefixed with "text.".
```

super small typo. I can fix in my PR.

junr03 commented:
@htuch @mattklein123 can either of you do a final pass to get the api-shepherds approval?

htuch commented on api/envoy/extensions/compression/common/compressor/v3/compressor.proto:
```diff
@@ -0,0 +1,22 @@
+message CompressorLibrary {
+  // Compressor library specific configuration. See the supported libraries for further
+  // documentation.
+  google.protobuf.Any typed_config = 1;
```
Can we move this one to the new `TypedConfig` that will appear in https://github.com/envoyproxy/envoy/pull/10908?

rojkov replied:
Sure, I've subscribed to that PR.

rojkov replied:
I presume it's about `TypedExtensionConfig`, not `TypedConfig`?

htuch replied:
Ack

junr03 replied:
@htuch If we are moving all the `typed_config`s across the project to the new `TypedExtenstionConfig` message that is being created in #10908, this could be updated then. I'd like to merge this and come back and update when the discussion in #10908 settles, and gets merged. We have been working on this PR for a while, and have a chain of a few other PRs that depend on this one.

htuch replied:
I don't think we can move them all without a deprecation and major version change. The idea is to make all new ones `TypedExtensionConfig`. Would it satisfy your concern if I submit a quick PR to introduce this independent of #10908?

htuch replied:
https://github.com/envoyproxy/envoy/pull/11105

junr03 replied:
> I don't think we can move them all without a deprecation and major version change. 

Yep, I was suggesting we would be doing the config in this PR as part of the deprecation that all other extensions would have to go through. But I see the want to wait a bit to land the correct thing that would avoid adding more volume to the deprecation cycle. Thanks for #11105, hopefully we can get that in soon, and move this PR along :)

junr03 commented on source/extensions/filters/http/compressor/compressor_filter.cc:
```diff
@@ -0,0 +1,26 @@
+    Compression::Compressor::CompressorFactoryPtr compressor_factory)
+    : Common::Compressors::CompressorFilterConfig(
+          generic_compressor,
+          stats_prefix + "compressor." + generic_compressor.stats_prefix() +
```
now that #11105 merged, we can delete the `stats_prefix` field, and use generic_compressor.compressor_library().typed_config().name().

rojkov commented:
Do we really need the `CompressorLibrary` message now when there is this standard Envoy-specific `TypedExtensionConfig`? I remember I had to introduce the new message to accommodate `config` and `typed_config`, but then we moved to `typed_config`s only and now the message looks redundant.

I've replaced `CompressorLibrary` with `TypedExtensionConfig` and the configuration of compressors still looks sane imho.

junr03 commented on api/envoy/extensions/compression/common/compressor/v3/compressor.proto:
```diff
@@ -1,22 +0,0 @@
-
-// [#protodoc-title: Compressor Library]
-
-message CompressorLibrary {
```
I guess the point of this message was to potentially have fields that were common to all compressor extensions. However, if we believe there won't be a case for this, I am happy to delete.

mattklein123 commented:
@rojkov  friendly request to please never force push as it makes reviews more difficult, thank you!
